Presence of 'pydantic' package breaks charmap()/charmap_file()

If `pydantic` is installed in the same venv as hypothesis, the `charmap()` logic seems to become broken. I can reproduce the problem using hypothesis' tests on git master (d8eeaf8000915cb5ad5d8eb3a1991b448dc4abd3), after adding pydantic to test requirements:

```
GLOB sdist-make: /tmp/hypothesis/hypothesis-python/setup.py
py37-brief inst-nodeps: /tmp/hypothesis/hypothesis-python/.tox/.tmp/package/1/hypothesis-6.7.0.zip
py37-brief installed: apipkg==1.5,attrs==20.3.0,execnet==1.8.0,hypothesis @ file:///tmp/hypothesis/hypothesis-python/.tox/.tmp/package/1/hypothesis-6.7.0.zip,importlib-metadata==3.7.2,importlib-resources==5.1.2,iniconfig==1.1.1,packaging==20.9,pexpect==4.8.0,pluggy==0.13.1,ptyprocess==0.7.0,py==1.10.0,pydantic==1.8.1,pyparsing==2.4.7,pytest==6.2.2,pytest-forked==1.3.0,pytest-xdist==2.2.1,sortedcontainers==2.3.0,toml==0.10.2,typing-extensions==3.7.4.3,zipp==3.4.1
py37-brief run-test-pre: PYTHONHASHSEED='908145994'
py37-brief run-test: commands[0] | python -m pytest tests/cover/test_testdecorators.py
ImportError while loading conftest '/tmp/hypothesis/hypothesis-python/tests/conftest.py'.
tests/conftest.py:28: in <module>
    run()
tests/common/setup.py:66: in run
    assert os.path.exists(charmap_file()), charmap_file()
E   AssertionError: /tmp/tmpnmoh0hu3/unicode_data/11.0.0/charmap.json.gz
ERROR: InvocationError for command /tmp/hypothesis/hypothesis-python/.tox/py37-brief/bin/python -m pytest tests/cover/test_testdecorators.py (exited with code 4)
____________________________________________________________________ summary _____________________________________________________________________
ERROR:   py37-brief: commands failed
```

I don't know if it's strictly related to `charmap_file()` or just another effect of pydantic but once I remove the assert, a number of tests fail (I've run the 'prettyquick' target), compared to without pydantic:

```
FAILED tests/cover/test_type_lookup.py::test_can_generate_from_all_registered_types[Color] - AssertionError: assert False
FAILED tests/cover/test_type_lookup.py::test_can_generate_from_all_registered_types[IPvAnyAddress] - AssertionError: assert False
FAILED tests/cover/test_type_lookup.py::test_can_generate_from_all_registered_types[IPvAnyInterface] - AssertionError: assert False
FAILED tests/cover/test_type_lookup.py::test_can_generate_from_all_registered_types[IPvAnyNetwork] - AssertionError: assert False
FAILED tests/cover/test_type_lookup.py::test_can_generate_from_all_registered_types[ConstrainedDecimal] - AssertionError: assert False
FAILED tests/cover/test_type_lookup.py::test_can_generate_from_all_registered_types[ConstrainedFloat] - AssertionError: assert False
FAILED tests/cover/test_type_lookup.py::test_can_generate_from_all_registered_types[ConstrainedInt] - AssertionError: assert False
FAILED tests/cover/test_type_lookup.py::test_can_generate_from_all_registered_types[Json] - AssertionError: assert False
FAILED tests/cover/test_type_lookup.py::test_can_generate_from_all_registered_types[NegativeFloat] - AssertionError: assert False
FAILED tests/cover/test_type_lookup.py::test_can_generate_from_all_registered_types[NegativeInt] - AssertionError: assert False
FAILED tests/cover/test_type_lookup.py::test_can_generate_from_all_registered_types[NonNegativeFloat] - AssertionError: assert False
FAILED tests/cover/test_type_lookup.py::test_can_generate_from_all_registered_types[NonNegativeInt] - AssertionError: assert False
FAILED tests/cover/test_type_lookup.py::test_can_generate_from_all_registered_types[NonPositiveFloat] - AssertionError: assert False
FAILED tests/cover/test_type_lookup.py::test_can_generate_from_all_registered_types[NonPositiveInt] - AssertionError: assert False
FAILED tests/cover/test_type_lookup.py::test_can_generate_from_all_registered_types[PaymentCardNumber] - AssertionError: assert False
FAILED tests/cover/test_type_lookup.py::test_can_generate_from_all_registered_types[PositiveFloat] - AssertionError: assert False
FAILED tests/cover/test_type_lookup.py::test_can_generate_from_all_registered_types[PositiveInt] - AssertionError: assert False
FAILED tests/cover/test_type_lookup.py::test_can_generate_from_all_registered_types[PyObject] - AssertionError: assert False
FAILED tests/cover/test_type_lookup.py::test_can_generate_from_all_registered_types[StrictBool] - AssertionError: assert False
FAILED tests/cover/test_type_lookup.py::test_can_generate_from_all_registered_types[StrictFloat] - AssertionError: assert False
FAILED tests/cover/test_type_lookup.py::test_can_generate_from_all_registered_types[StrictInt] - AssertionError: assert False
FAILED tests/cover/test_type_lookup.py::test_can_generate_from_all_registered_types[StrictStr] - AssertionError: assert False
FAILED tests/cover/test_type_lookup.py::test_can_generate_from_all_registered_types[UUID1] - AssertionError: assert False
FAILED tests/cover/test_type_lookup.py::test_can_generate_from_all_registered_types[UUID3] - AssertionError: assert False
FAILED tests/cover/test_type_lookup.py::test_can_generate_from_all_registered_types[UUID4] - AssertionError: assert False
FAILED tests/cover/test_type_lookup.py::test_can_generate_from_all_registered_types[UUID5] - AssertionError: assert False
```

--- a/hypothesis-python/tests/common/setup.py
+++ b/hypothesis-python/tests/common/setup.py
@@ -21,7 +21,7 @@ from hypothesis import Verbosity, settings
 from hypothesis._settings import not_set
 from hypothesis.configuration import set_hypothesis_home_dir
 from hypothesis.errors import NonInteractiveExampleWarning
-from hypothesis.internal.charmap import charmap, charmap_file
+from hypothesis.internal import charmap
 from hypothesis.internal.coverage import IN_COVERAGE_TESTS
 
 
@@ -62,8 +62,10 @@ def run():
     set_hypothesis_home_dir(new_home)
     assert settings.default.database.path.startswith(new_home)
 
-    charmap()
-    assert os.path.exists(charmap_file()), charmap_file()
+    # Remove the cache because we might have saved this before setting the new home dir
+    charmap._charmap = None
+    charmap.charmap()
+    assert os.path.exists(charmap.charmap_file()), charmap.charmap_file()
     assert isinstance(settings, type)
 
     # We do a smoke test here before we mess around with settings.

--- a/hypothesis-python/tests/cover/test_type_lookup.py
+++ b/hypothesis-python/tests/cover/test_type_lookup.py
@@ -200,8 +200,13 @@ def test_uninspectable_from_type():
         st.from_type(BrokenClass).example()
 
 
+def _check_instances(t):
+    # See https://github.com/samuelcolvin/pydantic/discussions/2508
+    return t.__module__ != "typing" and not t.__module__.startswith("pydantic")
+
+
 @pytest.mark.parametrize(
-    "typ", sorted((x for x in _global_type_lookup if x.__module__ != "typing"), key=str)
+    "typ", sorted((x for x in _global_type_lookup if _check_instances(x)), key=str)
 )
 @given(data=st.data())
 def test_can_generate_from_all_registered_types(data, typ):

